### DISCUSSION TITLE
Changing filename/path of uploaded file in a BEFORE INSERT trigger?

### DISCUSSION THREAD

**Discussion by wadclapp** [supabase/supabase#6087] [Questions] [answered]
```js
const { data, error } = await supabase
  .storage
  .from('xyz')
  .upload('the/path/something-string.jpg', file)
```

In a `BEFORE INSERT` trigger on `storage.objects` I've tried changing the filename (`NEW.name = REPLACE(NEW.name, 'something-string', 'something-else');`)

This puts the file in the correct folder, but the file when viewed in the dashboard has no mime type (and cannot be downloaded or viewed), and the data returned from the `upload()`'s `data` provides the wrong path.

```js
console.log(data) // { Key: 'bucket/the/path/something-string.jpg' }
```

Is this expected behaviour? Should I rethink what I'm doing here?

**Comment by GaryAustin1** [comment, marked as answer]
The storage api runs as a separate "task" outside of the database.  It does an insert to the database with the path you provided, when that insert completes it then does a request to storage (s3 right now) and uses the original file path to store the file.  It does not look at returned data from object insert to see if object name was changed.  Also after the file is stored in S3, the object table gets updated by original object name/path with the metadata including mime, so that will not get updated if you changed the name.

In general you should not mess with the storage tables.  I think a trigger in particular that changes the data, could cause conflict if Supabase added a trigger.

For now this renaming would need to be done client side, or thru your own api on a server.

You might put in a request in storage-api to have them look at returned object name, which technically could be done, if they considered it something safe to support.

**Reply by wadclapp** [reply]
Thanks Gary, that explains it, really helpful.

Problem I have is I'd like the user to provide an upload path like `the/path/id_placeholder/something_myvalue.jpg` and this get changed to `the/path/< nextval('my_id_sequence') >/something.jpg`. And reference the id in a table. If it was possible to include metadata with the image upload, I wouldn't need to change the filename, but I've been attempting to include in the string another value (`myvalue`) for a field to add in the same table.

> In general you should not mess with the storage tables. I think a trigger in particular that changes the data, could cause conflict if Supabase added a trigger.

I also `INSERT` a row in the database to accompany this file in the `BEFORE INSERT` trigger - how would I do that without creating a trigger on insert?

**Reply by GaryAustin1** [reply]
Having a trigger that does not change the object data, would mean it is safe from "hurting" anything Supabase might do.  Your risk is only then if Supabase changes the object table in a way that breaks your trigger.  For instance they changed the bucket table in a migration in the life of the project adding a field.  

Really not following your id sequence change stuff.  Can you use your other table to store the needed info and use that to generate the mapping to the file name?  But as it is,  you have to make any path/filename changes before you do the storage call to the API.
